### PR TITLE
Ed25519 async batch verification for JoinSplit signatures

### LINKED ISSUE #1944: Async batch Ed25519 signature verification
We currently validate Sprout JoinSplit binding signatures one by one. `ed25519-zebra` has batch validation math implemented, which `tower-batch` uses in an integration test to demonstrate  its use wrapped in a `tower::Service`. In fact this was the first test case for the pattern we've used for the async batch RedJubjub signature and Groth16 proof verification. We can use this in our transaction verification pipeline too to get the batch verification speedup.

### PR DESCRIPTION
**Pull request by dconnolly**
## Motivation

We've been verifying JoinSplitSigs one-by-one pre-ZIP-215. Now as we're post-ZIP-215,
we can take advantage of the batch math to validate this signatures.

## Solution

This adds another `tower::Service` that allows async batch validation of Ed25519 signatures. This is very similar boilerplate to `redjubjub::Verifier` and similar tests.
I removed the single check and replaced its invocation with a call to this communal `ed25519::Verifier` instance.

I would have pumped all the joinsplits in our MAINNET_BLOCKS test vectors but these
signatures are over the sighash, which needs the NU code to compute, and once we're
doing all that set up, we're basically doing transaction validation, so.

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Unit Tests and Property Tests

## Review

Not urgent, a quick chunk of work we've recently been unblocked on doing.

## Related Issues

Resolves #1944

## Follow Up Work

This is the third async batch verifier `tower::Service`, and won't be the last (Orchard will bring `halo2::Verifier` and `redpallas::Verifier` eventually). There is a lot of boilerplate being copy-pasted around, which may be ok for easy deletion, but when it's duplicated amongst ~5 variants, it's  a signal to revisit where we can DRY this up or make a more generic async batch Verifier, which can then be instantiated with different `Item`s/`ItemWrapper`s and some `call` method/function. Tracked in #1951.

These batch verifiers are returning their own errors from their underlying batchers, we should figure out the best way to turn those into `TransactionError`s, either in `zebra-consensus::transaction` or in the async service layer.

### CODE REVIEW

File: zebra-consensus/src/primitives/ed25519.rs
Comment (oxarbitrage) [inline zebra-consensus/src/primitives/ed25519.rs:5]:
```suggestion
```
Not needed.

File: zebra-consensus/src/transaction.rs
Comment (oxarbitrage) [inline zebra-consensus/src/transaction.rs]:
```suggestion
```
As the function `validate_joinsplit_sig` was removed from the codebase, lets remove the comment.

File: zebra-chain/Cargo.toml
Comment (oxarbitrage) [inline zebra-chain/Cargo.toml]: We should maybe follow up this and the `redjubjub` in a new ticket. I will prefer zebra dependencies to be stable.

File: zebra-consensus/src/primitives/ed25519.rs
Comment (oxarbitrage) [inline zebra-consensus/src/primitives/ed25519.rs]:
```suggestion
                tracing::trace!("got ed25519 item");
```

File: zebra-consensus/src/primitives/ed25519.rs
Comment (oxarbitrage) [inline zebra-consensus/src/primitives/ed25519.rs]:
```suggestion
                tracing::trace!("got ed25519 flush command");
```

**Review by oxarbitrage — CHANGES_REQUESTED** [review]
I am not an expert on async code but i think this looks good.

I found a few minors that we can fix before merging:

- remove some not needed code and commented code.
- add "ed25519" string to some tracing messages that are too generic and can be confused with other similar ones.

File: zebra-consensus/src/primitives/ed25519.rs
Comment (teor2345) [reply zebra-consensus/src/primitives/ed25519.rs:5]: This code imports the new `ed25519::tests` module when we're testing. Otherwise the tests don't run.

File: zebra-consensus/src/primitives/ed25519.rs
Comment (teor2345) [inline zebra-consensus/src/primitives/ed25519.rs:71]: I don't think this alias clarifies that at all 🤣

File: zebra-consensus/src/primitives/ed25519.rs
Comment (teor2345) [inline zebra-consensus/src/primitives/ed25519.rs]:
```suggestion
                                "ed25519 batch verification receiver lagged and lost verification results"
```

File: zebra-consensus/src/primitives/ed25519.rs
Comment (teor2345) [inline zebra-consensus/src/primitives/ed25519.rs]:
```suggestion
                        Err(RecvError::Closed) => panic!("ed25519 verifier was dropped without flushing"),
```

File: zebra-consensus/src/primitives/ed25519.rs
Comment (teor2345) [inline zebra-consensus/src/primitives/ed25519.rs:96]: Can you open a follow-up ticket so we add these metrics for all signature verification at all levels?

Ideally, the ticket should have a checklist of each signature type for transparent, Sprout, Sapling, and Orchard.

**Review by teor2345 — COMMENTED** [review]
Looks great!

This PR could merge as-is, but let's do some tweaks and open some follow-ups first.

**Comment by teor2345** [comment]
I'm testing some full mainnet and testnet syncs locally with this branch and #1950.

I don't want to block this PR on those tests, it's not a high risk PR.

File: zebra-consensus/src/primitives/ed25519.rs
Comment (oxarbitrage) [reply zebra-consensus/src/primitives/ed25519.rs:5]: Sorry about that, i misunderstood the code there. Thanks.

Comment (teor2345) [reply zebra-consensus/src/primitives/ed25519.rs:5]: No worries!

**Comment by teor2345** [comment]
> I'm testing some full mainnet and testnet syncs locally with this branch and #1950.

Seems to work fine after 2 full mainnet syncs and 2 full testnet syncs, and after syncing to tip.

**Comment by teor2345** [comment]
Seeing 2 mainnet instances at around 180% CPU, spread across all cores.

(The CPU might even be a bit elevated due to all the "Invalid binding signature" logs.)

File: zebra-chain/Cargo.toml
Comment (dconnolly) [reply zebra-chain/Cargo.toml]: https://github.com/ZcashFoundation/ed25519-zebra/issues/45

File: zebra-consensus/src/primitives/ed25519.rs
Comment (dconnolly) [reply zebra-consensus/src/primitives/ed25519.rs:96]: Done: https://github.com/ZcashFoundation/zebra/issues/1966

File: zebra-chain/Cargo.toml
Comment (teor2345) [reply zebra-chain/Cargo.toml]: Also added to https://github.com/ZcashFoundation/zebra/issues/1086

**Review by teor2345 — APPROVED** [review]
Looks great, let's do more batching!

**Comment by teor2345** [comment]
I'm going to admin-merge because @oxarbitrage's changes were addressed.

**Pull request merged**
